File: ios/ThermalGuard/Sources/Core/Building.swift
import Foundation

// 楼宇拓扑：8 层示意楼宇，A/B 两条楼梯贯通，1 楼大堂正门与 8 楼天台两个出口。
// 与 Web 端 building.js 保持同构，三端算出的路线必须一致。

struct BuildingNode: Hashable {
    enum Kind: Hashable { case stair, corridor, exit }

    let id: String
    let floor: Int
    let x: Double   // 0–100，向东为正
    let y: Double   // 0–100，向南为正
    let kind: Kind
    let stair: String?
}

struct BuildingEdge: Hashable {
    enum Kind: Hashable { case corridor, stair }

    let a: String
    let b: String
    let meters: Double
    let kind: Kind
    let stair: String?
}

enum Building {
    static let floorCount = 8

    private static let layout = (west: 10.0, mid: 50.0, east: 90.0, corridor: 48.0, lobby: 74.0, roof: 22.0)

    static let nodes: [String: BuildingNode] = {
        var nodes: [String: BuildingNode] = [:]
        for floor in 1...floorCount {
            nodes["A\(floor)"] = BuildingNode(id: "A\(floor)", floor: floor, x: layout.west, y: layout.corridor, kind: .stair, stair: "A")
            nodes["C\(floor)"] = BuildingNode(id: "C\(floor)", floor: floor, x: layout.mid, y: layout.corridor, kind: .corridor, stair: nil)
            nodes["B\(floor)"] = BuildingNode(id: "B\(floor)", floor: floor, x: layout.east, y: layout.corridor, kind: .stair, stair: "B")
        }
        nodes["L1"] = BuildingNode(id: "L1", floor: 1, x: layout.mid, y: layout.lobby, kind: .exit, stair: nil)
        nodes["ROOF"] = BuildingNode(id: "ROOF", floor: floorCount, x: layout.mid, y: layout.roof, kind: .exit, stair: nil)
        return nodes
    }()

    static let edges: [BuildingEdge] = {
        var edges: [BuildingEdge] = []
        for floor in 1...floorCount {
            edges.append(BuildingEdge(a: "A\(floor)", b: "C\(floor)", meters: 16, kind: .corridor, stair: nil))
            edges.append(BuildingEdge(a: "C\(floor)", b: "B\(floor)", meters: 16, kind: .corridor, stair: nil))
            if floor < floorCount {
                edges.append(BuildingEdge(a: "A\(floor)", b: "A\(floor + 1)", meters: 12, kind: .stair, stair: "A"))
                edges.append(BuildingEdge(a: "B\(floor)", b: "B\(floor + 1)", meters: 12, kind: .stair, stair: "B"))
            }
        }
        edges.append(BuildingEdge(a: "C1", b: "L1", meters: 8, kind: .corridor, stair: nil))
        edges.append(BuildingEdge(a: "C\(floorCount)", b: "ROOF", meters: 10, kind: .corridor, stair: nil))
        return edges
    }()

    static let adjacency: [String: [BuildingEdge]] = {
        var map: [String: [BuildingEdge]] = [:]
        for id in nodes.keys { map[id] = [] }
        for edge in edges {
            map[edge.a, default: []].append(edge)
            map[edge.b, default: []].append(edge)
        }
        return map
    }()

    /// 出口与其附加代价：天台只在向下通道不可通行时才会被选中
    static let exits: [(id: String, label: String, penalty: Double)] = [
        (id: "L1", label: "1 楼大堂正门", penalty: 0),
        (id: "ROOF", label: "8 楼天台避难区", penalty: 400),
    ]

    static func node(_ id: String) -> BuildingNode? { nodes[id] }

    static func label(_ id: String) -> String {
        switch id {
        case "L1": return "1 楼大堂正门"
        case "ROOF": return "8 楼天台避难区"
        default:
            guard let node = nodes[id] else { return id }
            let spot = node.kind == .stair ? "\(node.stair ?? "") 楼梯口" : "走廊中段"
            return "\(node.floor) 楼\(spot)"
        }
    }

    static func positionNodeId(floor: Int, spot: String) -> String { "\(spot)\(floor)" }

    static func neighbors(_ id: String) -> [BuildingEdge] { adjacency[id] ?? [] }

    static func edge(from: String, to: String) -> BuildingEdge? {
        neighbors(from).first { $0.a == to || $0.b == to }
    }
}

File: src/mobile/building.js
// 楼宇拓扑模型（8 层高密度旧式住宅示意结构）
// 每层：A 楼梯口（西）— 走廊中段 — B 楼梯口（东）
// 垂直：A / B 两条楼梯贯通 1–8 楼
// 出口：1 楼大堂正门、8 楼天台避难区
//
// 坐标使用 0–100 的归一化平面坐标，x 向右为东，y 向下为南。
// 距离单位统一为米，便于直接换算步行时间。

export const FLOOR_COUNT = 8

export const STAIRS = {
  A: { id: 'A', short: 'A 楼梯', label: 'A 楼梯（西侧）' },
  B: { id: 'B', short: 'B 楼梯', label: 'B 楼梯（东侧）' },
}

export const EXIT_LIST = [
  { id: 'L1', label: '1 楼大堂正门', short: '正门', penalty: 0 },
  // 消防原则是优先向下撤离至地面，天台只作为向下通道不可通行（或代价极高）时的备选，
  // 因此给天台附加一个很高的代价（约等于连穿 7 段浓烟的惩罚）。
  { id: 'ROOF', label: '8 楼天台避难区', short: '天台', penalty: 400 },
]

export const LAYOUT = { west: 10, mid: 50, east: 90, corridor: 48, lobby: 74, roof: 22 }

function buildGraph() {
  const nodes = {}
  const edges = []

  const link = (a, b, meters, kind, extra = {}) => {
    edges.push({ a, b, meters, kind, ...extra })
  }

  for (let floor = 1; floor <= FLOOR_COUNT; floor += 1) {
    nodes[`A${floor}`] = { id: `A${floor}`, floor, x: LAYOUT.west, y: LAYOUT.corridor, kind: 'stair', stair: 'A' }
    nodes[`C${floor}`] = { id: `C${floor}`, floor, x: LAYOUT.mid, y: LAYOUT.corridor, kind: 'corridor' }
    nodes[`B${floor}`] = { id: `B${floor}`, floor, x: LAYOUT.east, y: LAYOUT.corridor, kind: 'stair', stair: 'B' }

    link(`A${floor}`, `C${floor}`, 16, 'corridor')
    link(`C${floor}`, `B${floor}`, 16, 'corridor')

    if (floor < FLOOR_COUNT) {
      link(`A${floor}`, `A${floor + 1}`, 12, 'stair', { stair: 'A' })
      link(`B${floor}`, `B${floor + 1}`, 12, 'stair', { stair: 'B' })
    }
  }

  nodes.L1 = { id: 'L1', floor: 1, x: LAYOUT.mid, y: LAYOUT.lobby, kind: 'exit', exitId: 'L1' }
  link('C1', 'L1', 8, 'corridor', { exitId: 'L1' })

  nodes.ROOF = { id: 'ROOF', floor: FLOOR_COUNT, x: LAYOUT.mid, y: LAYOUT.roof, kind: 'exit', exitId: 'ROOF' }
  link(`C${FLOOR_COUNT}`, 'ROOF', 10, 'corridor', { exitId: 'ROOF' })

  const adjacency = {}
  Object.keys(nodes).forEach((id) => { adjacency[id] = [] })
  edges.forEach((edge) => {
    adjacency[edge.a].push({ ...edge, to: edge.b })
    adjacency[edge.b].push({ ...edge, to: edge.a })
  })

  return { nodes, edges, adjacency }
}

export const BUILDING = buildGraph()

export function nodeById(id) {
  return BUILDING.nodes[id]
}

export function floorNodeIds(floor) {
  return [`A${floor}`, `C${floor}`, `B${floor}`]
}

export function nodeLabel(id) {
  const node = BUILDING.nodes[id]
  if (!node) return id
  if (id === 'L1') return '1 楼大堂正门'
  if (id === 'ROOF') return '8 楼天台避难区'
  const spot = node.kind === 'stair' ? `${node.stair} 楼梯口` : '走廊中段'
  return `${node.floor} 楼${spot}`
}

export function nodeShortLabel(id) {
  const node = BUILDING.nodes[id]
  if (!node) return id
  if (id === 'L1') return '正门'
  if (id === 'ROOF') return '天台'
  return node.kind === 'stair' ? `${node.stair} 梯` : '走廊'
}

export function positionNodeId(floor, spot) {
  return `${spot}${floor}`
}
